src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,132 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private class Node<T> {
+        private Node<T> prev;
+        private Node<T> next;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.next = next;
+            this.value = value;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<T>(tail, value, null);
+        if (head == null) {
+            head = tail = newNode;
+        }
+        tail.next = newNode;
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size && size == 0) {
+            add(value);
+        }
+        validateIndex(index);
+        if (size == index) {
+            add(value);
+        }
+        Node<T> currentNode = nodeByIndex(index);;
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        currentNode.prev = newNode;
+        if (index == 0) {
+            head = newNode;
+        } else {
+            newNode.prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
```

Comment:
We can use for each loop to go through the list.

Reply:
why should i if i need a counter?

Reply:
You can, but it's not necessary.